atlasdb-service/src/main/java/com/palantir/atlasdb/jackson/AtlasDeserializers.java

```diff
@@ -79,16 +79,24 @@ private AtlasDeserializers() {
                 "Received %s values for a row with %s components.", size, components.size());
         byte[][] bytes = new byte[size][];
         for (int i = 0; i < size; i++) {
-            JsonNode value = node.get(i);
             NameComponentDescription component = components.get(i);
-            bytes[i] = component.getType().convertFromJson(value.toString());
+            String value = getRowValue(i, component, node);
+            bytes[i] = component.getType().convertFromJson(value);
             if (component.isReverseOrder()) {
                 EncodingUtils.flipAllBitsInPlace(bytes[i]);
             }
         }
         return Bytes.concat(bytes);
     }
 
+    private static String getRowValue(int i, NameComponentDescription component, JsonNode node) {
+        if (node.has(component.getComponentName())) {
```

Comment:
It is a bit strange to do this check for every field. The logic should probably be like
```
private static Iterator<JsonNode> getComponentNodes(JsonNode node, List<NameComponentDescription> components) {
    if (node.isArray()) {
        return node.elements(); // iterator over array's elements  
    } else {
        // Lazily transform the list of components to an iterable of child nodes
        return Iterators.transform(components.iterator(), c -> node.get(component.getComponentName()));       
    }
}
```